dev-tool/src/list_net/data_source.rs

```diff
@@ -0,0 +1,567 @@
+
+    /// Loads a batch of samples and returns a vector of references to them.
+    ///
+    /// See [`Storage.load_batch()`] about why this is `&mut self`.
+    fn load_batch<'a>(&'a mut self, ids: &'_ [DataId]) -> Result<Vec<Sample<'a>>, Self::Error>;
+}
+
+/// For now samples id's are always incremental integers form `0..nr_samples`.
+///
+/// Given that this is just used to handle shuffling and similar it's unlikely to
+/// ever change, we do explicitly not care about any consistency of the id to
+/// sample mapping between trainings.
+type DataId = usize;
+
+/// Helper to randomize sample order.
+struct DataLookupOrder {
+    /// Ids of all samples in this lookup order.
+    data_ids: Vec<DataId>,
+    /// The offset splitting already used samples from not yet used samples.
+    data_ids_offset: usize,
+}
+
+impl DataLookupOrder {
+    fn new(data_ids: Vec<DataId>) -> Self {
+        Self {
+            data_ids,
+            data_ids_offset: 0,
+        }
+    }
+
+    /// Returns the number of batches which will be produced with given batch size.
+    fn number_of_batches(&self, batch_size: usize) -> usize {
+        self.data_ids.len() / batch_size
+    }
+
+    /// Resets this container.
+    ///
+    /// This will mark all samples as unused and
+    /// shuffles the order of samples.
+    fn reset(&mut self, rng: &mut impl Rng) {
+        self.data_ids.shuffle(rng);
+        self.data_ids_offset = 0;
+    }
+
+    /// Returns the next unused sample id.
+    fn next(&mut self) -> Option<DataId> {
+        if self.data_ids_offset < self.data_ids.len() {
+            let idx = self.data_ids_offset;
+            self.data_ids_offset += 1;
+
+            Some(self.data_ids[idx])
+        } else {
+            None
+        }
+    }
+
+    /// Returns the next `batch_size` many sample ids.
+    fn next_batch(&mut self, batch_size: usize) -> Vec<DataId> {
+        let end_idx = self.data_ids_offset + batch_size;
+        if end_idx <= self.data_ids.len() {
+            let start_idx = self.data_ids_offset;
+            self.data_ids_offset = end_idx;
+
+            self.data_ids[start_idx..end_idx].to_owned()
+        } else {
+            Vec::new()
+        }
+    }
+}
+
+/// A "in-memory" sample storage.
+///
+/// It also can be portably stored to disk using bincode serialization.
+///
+/// The storage is reasonably compact as ~5.3GiB of soundgarden user
+/// data-frames will produce a less then 975MiB serialized `.samples` file.
+// While there are many ways to improve on this (e.g. memory-mapped I/O), they are not relevant for our
+// use-case for now.
+#[derive(Serialize, Deserialize, Default)]
+pub(crate) struct InMemorySamples {
+    /// Vector of concatenated `inputs` and `target_prob_dist`, this slightly improves memory size and
+    /// cache locality, we can derive the number of document from the vectors length as the vectors length
+    /// is `nr_document * INPUT_NR_FEATURES + nr_document*1`, i.e. `nr_documents * 51`.
```

Comment:
can you explain how "this slightly improves memory size and cache locality"?

Reply:
The alternative would be `Vec<(Vec<f32>, Vec<f32>)>` (or `Vec<(Array2<f32>, Array1<f32>)>` which has a bit more overhead due to storing a `Vec<f32>` and a `Ix2`, not that it matters with the data sizes we did got, through it might have made a difference with the sizes I expected to have).

So the sample features and target probability distribution might be in very different parts of the heap and in turn cause more cache misses.

I.e. given a cache line size of 64 (through modern processors have 128 or double fetch) the samples cover 31.25 cache lines and the target prob distribution covers 0.625 cache lines. So by storing them "inline" we will use 32 cache lines while when storing them separate we most likely will use 33 cache lines. (In case of modern CPUs it's more like 15 vs. 16 cache fetches). The memory/storage size improvement s are negligible, I guess.

Through that comment above is actually more though then I put into it when implementing it. 

I just thought it would be good to keep things together in memory and expected much larger data sources then what we have in the end. Then after implementing it (without much time investment) I realized that it was a pre-mature optimization but thought that it also shouldn't hurt to keep it.

So if you insist I can remove it.

Reply:
thanks, no need to remove it, just that I found it hard to understand the comment w/o knowing what the "alternative" was that you had in mind. perhaps you can add some of the above e.g. the first sentence into the comment but I'll leave that to your judgement.